app/javascript/components/explore/plotDataCache.js

```diff
@@ -0,0 +1,278 @@
+          data: {},
+          isPureCache: true // set a flag indicating that no fresh request to the server was needed
+        }, 0 // a backend time of zero is appropriate since this request will never go to the server
+      ])
+    }
+    promises.push(apiCallPromise)
+
+    return Promise.all(promises).then(resultArray => {
+      let mergedResult = null
+      resultArray.forEach(result => {
+        mergedResult = cache._mergeClusterResponse(studyAccession, result, cluster, annotation, subsample)
+      })
+      return mergedResult
+    })
+  }
+
+  /** Wipes the entire cache */
+  cache.clear = () => {
+    cache.entries = {}
+  }
+
+
+  /** adds the data for a given study/clusterName, overwriting any previous entry */
+  cache._mergeClusterResponse = (accession, clusterResponse, requestedCluster, requestedAnnotation, requestedSubsample) => {
+    const scatter = clusterResponse[0]
+    const cacheEntry = cache._findOrCreateEntry(accession, scatter.cluster, scatter.subsample)
+
+    if (scatter.cluster != requestedCluster || requestedSubsample !== scatter.subsample) {
+      // if the returned cluster name is different (likely because we requested '_default' and then
+      // got back the actual cluster name), also cache the response under the name of the requested name
+      cache._putEntry(accession, requestedCluster, requestedSubsample, cacheEntry)
+    }
+    if (scatter.isPureCache) {
+      // we need the response cluster/subsample to mimic what actually came from the server for the graphs
+      // to render correctly
+      scatter.annotParams = cacheEntry.clusterProps.annotParams
+      scatter.cluster = cacheEntry.clusterProps.cluster
+      scatter.subsample = cacheEntry.clusterProps.subsample
+    }
+    FIELDS.clusterProps.merge(cacheEntry, scatter)
+    FIELDS.cellsAndCoords.merge(cacheEntry, scatter)
+    FIELDS.annotation.merge(cacheEntry, scatter)
+
+    if (scatter.genes.length) {
+      FIELDS.expression.merge(cacheEntry, scatter)
+    }
+    return clusterResponse
+  }
+
+  /** get the data for a given study/cluster.  Returns a blank entry if none exists */
+  cache._findOrCreateEntry = (accession, clusterName, subsample) => {
+    const key = getKey(accession, clusterName, subsample)
+
+    if (!cache.entries[key]) {
+      cache.entries[key] = {
+        clusterProps: {},
+        cellsAndCoords: {},
+        annotations: {},
+        expression: {}
+      }
+    }
+    return cache.entries[key]
+  }
+
+  /** get the data for a given study/cluster */
+  cache._putEntry = (accession, clusterName, subsample, entry) => {
+    const key = getKey(accession, clusterName, subsample)
+    cache.entries[key] = entry
+  }
+
+  /** based on cache contents and desired return values, returns an array suitable for the 'fields' argument
+    * to fetchCluster in scp-api, and an array of promises for any in-flight requests relating to the needed data
+     */
+  cache._getFieldsToRequest = ({
```

Comment:
The structure of this method seems overly complicated... I'm wondering if it can be split up into some smaller helper methods to reduce the cyclomatic complexity.  I'm having a hard time reading it with all of the nested if/else logic.